File: Fabric/src/main/java/gay/object/hexdebug/fabric/HexDebugAbstractionsImpl.java
package gay.object.hexdebug.fabric;

import gay.object.hexdebug.IHexDebugAbstractions;
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayConnectionEvents;
import net.fabricmc.fabric.api.event.lifecycle.v1.ServerLifecycleEvents;
import net.minecraft.client.Minecraft;
import net.minecraft.server.MinecraftServer;

import java.util.function.Consumer;

public class HexDebugAbstractionsImpl implements IHexDebugAbstractions {
    public static IHexDebugAbstractions get() {
        return new HexDebugAbstractionsImpl();
    }

    @Override
    public void initPlatformSpecific() {
        HexDebugConfigFabric.init();
    }

    @Override
    public void onServerStarted(Consumer<MinecraftServer> callback) {
        ServerLifecycleEvents.SERVER_STARTED.register(callback::accept);
    }

    @Override
    public void onServerStopping(Consumer<MinecraftServer> callback) {
        ServerLifecycleEvents.SERVER_STOPPING.register(callback::accept);
    }

    @Override
    public void onClientJoin(Consumer<Minecraft> callback) {
        ClientPlayConnectionEvents.JOIN.register((l, s, client) -> callback.accept(client));
    }

    @Override
    public void onClientDisconnect(Consumer<Minecraft> callback) {
        ClientPlayConnectionEvents.DISCONNECT.register((l, client) -> callback.accept(client));
    }
}

File: Fabric/src/main/kotlin/gay/object/hexdebug/fabric/HexDebugAbstractionsImpl.kt
@file:JvmName("HexDebugAbstractionsImpl")

package gay.`object`.hexdebug.fabric

import gay.`object`.hexdebug.registry.HexDebugRegistrar
import net.fabricmc.fabric.api.resource.IdentifiableResourceReloadListener
import net.fabricmc.fabric.api.resource.ResourceManagerHelper
import net.minecraft.core.Registry
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.packs.PackType
import net.minecraft.server.packs.resources.PreparableReloadListener
import net.minecraft.server.packs.resources.ResourceManager
import net.minecraft.util.profiling.ProfilerFiller
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executor

fun <T : Any> initRegistry(registrar: HexDebugRegistrar<T>) {
    val registry = registrar.registry
    registrar.init { id, value -> Registry.register(registry, id, value) }
}

fun registerClientResourceReloadListener(id: ResourceLocation, listener: PreparableReloadListener) {
    ResourceManagerHelper.get(PackType.CLIENT_RESOURCES).registerReloadListener(
        object : IdentifiableResourceReloadListener {
            override fun reload(
                preparationBarrier: PreparableReloadListener.PreparationBarrier,
                resourceManager: ResourceManager,
                preparationsProfiler: ProfilerFiller,
                reloadProfiler: ProfilerFiller,
                backgroundExecutor: Executor,
                gameExecutor: Executor
            ): CompletableFuture<Void> {
                return listener.reload(preparationBarrier, resourceManager, preparationsProfiler, reloadProfiler, backgroundExecutor, gameExecutor)
            }

            override fun getFabricId() = id
        }
    )
}
